### PR TITLE
Setup scriv and adjust release scripts

### PR DESCRIPTION
**Pull request by michaelpj**
This sets up `scriv` to be used in the repo.

- Add `scriv` to the devshell and setup `changelog.d` folders.
- Worked on the release scripts. Now there are three:
    - `update-version.sh` updates the version for a package
    - `assemble-changelog.sh` assembles the changelog for a package
    - `prepare-release.sh` calls the above two for a list of packages
- Documented the above in the release doc and CONTRIBUTING

### CODE REVIEW

**Comment by zeme-iohk** [comment]
This look great!
Would be nice to turn `prepare-release.sh update-version.sh assembre-changelog.sh` into derivations and create devshell commands? 
I kinda want to get rid the `scripts` folder altogether (and it's gonna happen now that we are moving benchmarking to cicero)

**Comment by michaelpj** [comment]
> Would be nice to turn prepare-release.sh update-version.sh assembre-changelog.sh into derivations and create devshell commands?

I'm not sure how I feel about that... I mean, it doesn't really add anything (except discoverability) because they don't need to be built, and means you can't use them from outside nix. I guess I continue to want the ability to add things to the devshell menu without having to provide the command as a derivation right there...

**Comment by zeme-iohk** [comment]
> I'm not sure how I feel about that... I mean, it doesn't really add anything (except discoverability) because they don't need to be built

Sure, but l think that discoverability might be enough of a reason to do this.. 
As well as not having a ./scripts folder with random & unrelated stuff in it.

>and means you can't use them from outside nix.

You already can't use them outside nix, since they rely on `scriv`, which we provision through nix.

> I guess I continue to want the ability to add things to the devshell menu without having to provide the command as a derivation right there...

Mm we could do something like this:

```
  commands = [
    {
      package = cell.library.make-devshell-command ./scripts/update-release.sh 
      ...
    }
```
Where `cell.library.make-devshell-command` should be easy enough to implement.

But then we are back to the discussion of implicit runtime dependencies:
In this case `./scripts/update-release.sh` relies on `scriv` which is defined in the devshell, and if we removed it would break the script.

**Comment by michaelpj** [comment]
> You already can't use them outside nix, since they rely on scriv, which we provision through nix.

You can install `scriv` yourself, which is easy enough :shrug: 

I do agree that I'm being inconsistent here about whether or not I care about implicit runtime dependencies :D IDK. And I guess I'm okay with `fix-stylish-haskell` and co being only available via nix, so :shrug:

**Comment by zeme-iohk** [comment]
It just would've been nice to see them in the devshell menu, that's all

**Review by zeme-iohk — APPROVED** [review]

**Comment by michaelpj** [comment]
Folks who haven't commented, any thoughts?

**Review by thealmarty — APPROVED** [review]

**Comment by michaelpj** [comment]
> So it's not for every PR? It's a case by case basis? We use our own judgment if the PR merits a change fragment or not?

Great point! I should definitely mention this.

**Review by zeme-iohk — APPROVED** [review]

**Comment by michaelpj** [comment]
@zliu41 ^^

File: scripts/update-version.sh [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  echo "$USAGE"
+fi
+
+set -euo pipefail
+
+PACKAGE=$1
+VERSION=$2
+
+versioned_packages_regex=$(printf "\|%s" "${versioned_packages[@]}")
+versioned_packages_regex="\(${versioned_packages_regex:2}\)"
```
Comment (zliu41) [inline scripts/update-version.sh:20]: Not needed

**Review by zliu41 — COMMENTED** [review]
Nice. Does this complete https://input-output.atlassian.net/browse/PLT-980?

**Comment by michaelpj** [comment]
Last call for thoughts, @kwxm and @bezirg

File: CONTRIBUTING.adoc [hunk 2]
```diff
@@ -364,6 +364,50 @@ Set it up if you can, it's relatively easy to do.
...
+==== How to write a changelog entry
+
+The basic idea is that you write a changelog "fragment" in the `changelog.d` directory.
+When we do a release, these will be collected into the main `CHANGELOG.md`.
+Usually we don't edit `CHANGELOG.md` directly.
+
+You can make a changelog fragment using `scriv create` in the package directory, but you can also just create the fragment directly with an editor.
+A fragment is a markdown file beginning with a header giving the category of change.
+Currently these are:
+- Removed
```
Comment (kwxm) [inline CONTRIBUTING.adoc:389]: Are these supposed to be displayed as an itemised list?  I just see
```
A fragment is a markdown file beginning with a header giving the category of change. 
Currently these are: - Removed - Added - Changed - Deprecated - Fixed - Security
```

File: scripts/prepare-release.sh [hunk 3]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
+
+# for some reason you can't just put a heredoc in a variable...
```
Comment (kwxm) [inline scripts/prepare-release.sh:3]: You could make this into a shell function, like
```
usage () {
  echo "$(basename $0) VERSION [PACKAGE...]
  Prepares to release PACKAGEs at VERSION. If no PACKAGEs are provided,
  prepares to release the default packages."
}
```

and then just say `usage` later.

File: scripts/prepare-release.sh [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+read -r -d '' USAGE << EOF
+prepare-release.sh VERSION [PACKAGE...]
```
Comment (kwxm) [inline scripts/prepare-release.sh:5]:
```suggestion
$(basename $0) VERSION [PACKAGE...]
```

File: scripts/update-version.sh [hunk 5]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
+
+# for some reason you can't just put a heredoc in a variable...
```
Comment (kwxm) [inline scripts/update-version.sh:3]: Again, you could use a shell function here.

File: scripts/update-version.sh [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+read -r -d '' USAGE << EOF
+update-version.sh PACKAGE VERSION
```
Comment (kwxm) [inline scripts/update-version.sh:5]:
```suggestion
$(basename $0) PACKAGE VERSION
```

File: scripts/assemble-changelog.sh [hunk 7]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+
+# for some reason you can't just put a heredoc in a variable...
```
Comment (kwxm) [inline scripts/assemble-changelog.sh:3]: See later comments.

**Review by bezirg — APPROVED** [review]

File: scripts/update-version.sh [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+# It looks for patterns like the following:
+#
+# - ", plutus-core"
+# - ", plutus-core:uplc"
+# - ", plutus-core ^>=1.0"
+# - ", plutus-core:{plutus-core, plutus-core-testlib}  ^>=1.0"
+#
+# and updates the version bounds to "^>={major version}"
+echo "Updating version bounds on $PACKAGE to '>=$major_version'"
+find . -name "*.cabal" -exec sed -i "s/\(, $PACKAGE[^^]*\).*/\1 ^>=$major_version/" {} \;
```
Comment (kwxm) [inline scripts/update-version.sh:37]: A courageous use of `sed -i` there.

File: scripts/prepare-release.sh [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  Prepares to release PACKAGEs at VERSION. If no PACKAGEs are provided,
+  prepares to release the default packges.
```
Comment (kwxm) [inline scripts/prepare-release.sh:7]:
```suggestion
  prepares to release the default packages.
```

**Review by kwxm — APPROVED** [review]
Looks fine.  Let's see how it works out in practice.

File: scripts/assemble-changelog.sh [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+read -r -d '' USAGE << EOF
+assemble-changelog.sh PACKAGE VERSION
+  Assembles the changelog for PACKAGE at VERSION.
+EOF
+
+if [ "$#" == "0" ]; then
+  echo "$USAGE"
+fi
```
Comment (kwxm) [inline scripts/assemble-changelog.sh:11]: Oh wait.  Don't you need `exit 1` or something here?  I think that just prints out the message and keeps going.  It's similar in the other scripts.

**Comment by michaelpj** [comment]
Thanks Kenneth, all good comments, applied.

**Pull request merged**